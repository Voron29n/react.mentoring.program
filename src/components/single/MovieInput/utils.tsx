import React from 'react';
import { InputAdornment, TextField } from '@mui/material';
import { FormikValues } from 'formik';
import { FieldMetaProps } from 'formik/dist/types';
import { GenreCheckmarks, MovieKey } from 'components';
import { GENRE_LIST, mapRuntimeToString } from 'utils';

export const initComponentUtil =
  (
    getFieldMeta: <Value>(name: string) => FieldMetaProps<Value>,
    handleGenresChange: (genres: string[]) => void,
    handleFocusRuntime: () => void,
    handleBlurRuntime: () => void,
    handleChange: (e: React.ChangeEvent<any>) => void
  ) =>
  (
    keyName: MovieKey,
    placeholder: string,
    values: FormikValues,
    isRuntimeFocus: boolean
  ) => {
    const meta = getFieldMeta(keyName);
    const commonProps = {
      name: keyName,
      id: keyName,
      placeholder
    };

    switch (keyName) {
      case MovieKey.GENRE:
        if (!Array.isArray(values[keyName])) {
          return null;
        }
        return (
          <GenreCheckmarks
            meta={meta}
            value={values.genres}
            onChange={handleGenresChange}
            genreList={GENRE_LIST}
            {...commonProps}
          />
        );
      case MovieKey.RUNTIME:
        return (
          <TextField
            onFocus={handleFocusRuntime}
            onBlur={handleBlurRuntime}
            onChange={handleChange}
            error={Boolean(meta.error)}
            helperText={meta.error}
            value={
              isRuntimeFocus
                ? values[keyName]
                : mapRuntimeToString(values[keyName])
            }
            type={isRuntimeFocus ? 'number' : 'text'}
            InputProps={
              isRuntimeFocus
                ? {
                    endAdornment: (
                      <InputAdornment position='start'>min</InputAdornment>
                    )
                  }
                : null
            }
            {...commonProps}
          />
        );
      default:
        const type =
          keyName === MovieKey.DATE
            ? 'date'
            : keyName === MovieKey.RATING
            ? 'number'
            : 'text';

        return (
          <TextField
            multiline={keyName === MovieKey.OVERVIEW}
            rows={keyName === MovieKey.OVERVIEW ? 7 : 1}
            type={type}
            onChange={handleChange}
            error={Boolean(meta.error)}
            helperText={meta.error}
            value={values[keyName]}
            {...commonProps}
          />
        );
    }
  };
