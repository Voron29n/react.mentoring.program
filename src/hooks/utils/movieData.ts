import { Movie } from 'components';

export interface ServerMovie {
  id: number;
  title: string;
  genres: string[];
  release_date: string;
  vote_average: number;
  poster_path: string;
  runtime: number;
  overview: string;

  [key: string]: string | number | Array<string>;
}

export const isServerMovie = (
  movie: Movie | ServerMovie
): movie is ServerMovie =>
  (movie as ServerMovie).poster_path !== undefined &&
  (movie as ServerMovie) !== undefined;

const commonRestFields = ({
  id,
  title,
  overview,
  runtime,
  genres,
  ...rest
}: Movie | ServerMovie) => ({
  common: { id, title, overview, runtime, genres },
  rest: rest
});

export function transformMovieMap<T extends Movie | ServerMovie>(
  movie: Movie | ServerMovie
): T {
  const { common, rest } = commonRestFields(movie);
  return isServerMovie(movie)
    ? ({
        ...common,
        releaseDate: rest.release_date,
        voteAverage: rest.vote_average,
        posterPath: rest.poster_path
      } as T)
    : ({
        ...common,
        release_date: rest.releaseDate,
        vote_average: rest.voteAverage,
        poster_path: rest.posterPath
      } as T);
}
