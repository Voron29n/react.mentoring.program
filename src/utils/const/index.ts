export * from './buttons';
export * from './form';
export * from './viewComponents';
export * from './api';
