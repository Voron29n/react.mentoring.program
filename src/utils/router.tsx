import React from 'react';
import { createBrowserRouter } from 'react-router-dom';
import { Redirect, SearchMovie, UnknownPage } from 'components';

export type SearchParams = {
  searchQuery: string;
};

export const router = createBrowserRouter([
  {
    path: '/',
    element: <Redirect redirectPath={'search'} />,
    errorElement: <UnknownPage />
  },
  {
    path: '/search',
    element: <SearchMovie />,
    children: [
      {
        path: ':searchQuery'
      }
    ]
  }
]);
