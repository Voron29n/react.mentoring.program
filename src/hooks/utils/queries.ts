import { baseApiConfig } from 'utils';

export enum HttpMethod {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  DELETE = 'DELETE'
}

type IFetchProps = {
  url: string;
  method?: HttpMethod;
  formData?: FormData | string;
  headers?: HeadersInit;
};

export const fetchReq = async ({
  url,
  method = HttpMethod.GET,
  formData,
  headers
}: IFetchProps): Promise<Response> => {
  const res = await fetch(url, {
    method,
    body: formData,
    headers
  });
  if (!res.ok) {
    throw new Error(`Could not fetch ${url}, status: ${res.status}`);
  }
  return res;
};

enum SearchBy {
  GENRES = 'genres',
  TITLE = 'title'
}

export const generateMoviesUrl = (
  activeGenre: string,
  activeSortType: string,
  searchQuery: string
): URL => {
  const url = new URL(
    `${baseApiConfig._apiBase}movies?limit=${baseApiConfig._baseLimit}&offset=${baseApiConfig._baseOffset}`
  );
  if (activeSortType) {
    url.searchParams.append(baseApiConfig._searchParams.sortBy, activeSortType);
    url.searchParams.append(
      baseApiConfig._searchParams.sortOrder,
      baseApiConfig._baseSortOrder
    );
  }
  if (activeGenre) {
    url.searchParams.append(baseApiConfig._searchParams.filter, activeGenre);
  }
  if (searchQuery) {
    url.searchParams.append(
      baseApiConfig._searchParams.searchQuery,
      searchQuery
    );
    url.searchParams.append(
      baseApiConfig._searchParams.searchBy,
      SearchBy.TITLE
    );
  }
  return url;
};
