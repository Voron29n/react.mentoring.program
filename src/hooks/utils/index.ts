import { isServerMovie, ServerMovie, transformMovieMap } from './movieData';
import { fetchReq, generateMoviesUrl, HttpMethod } from './queries';

export {
  isServerMovie,
  transformMovieMap,
  fetchReq,
  generateMoviesUrl,
  ServerMovie,
  HttpMethod
};
