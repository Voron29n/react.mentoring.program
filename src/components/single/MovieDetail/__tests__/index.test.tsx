import React from 'react';
import { render, RenderResult, screen } from '@testing-library/react';
import { mapRuntimeToString } from 'utils';
import { MovieDetail } from '../index';
import { testMovie } from './mocks';

describe('MovieDetail component', () => {
  let movieDetailRender: RenderResult;

  beforeEach(() => {
    movieDetailRender = render(<MovieDetail movieDetail={testMovie} />);
  });

  it('renders', () => {
    const genres = testMovie.genres.join(', ');
    const releaseYear = new Date(testMovie.releaseDate).getFullYear();
    const runtime = mapRuntimeToString(testMovie.runtime);

    expect(screen.getByRole('img')).toHaveAttribute(
      'src',
      testMovie.posterPath
    );
    expect(screen.getByRole('img')).toHaveAttribute('alt', testMovie.title);
    expect(screen.getByText(testMovie.title)).toBeInTheDocument();
    expect(screen.getByText(testMovie.voteAverage)).toBeInTheDocument();
    expect(screen.getByText(genres)).toBeInTheDocument();
    expect(screen.getByText(releaseYear)).toBeInTheDocument();
    expect(screen.getByText(runtime)).toBeInTheDocument();
  });

  it('MovieDetail snapshot', () => {
    expect(movieDetailRender).toMatchSnapshot();
  });
});
