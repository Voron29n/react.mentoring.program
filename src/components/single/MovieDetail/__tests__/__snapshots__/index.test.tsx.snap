// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetail component MovieDetail snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="movie__detail__container"
      >
        <img
          alt="testTitle"
          class="poster__img"
          src="testPosterPath"
        />
        <div
          class="info"
        >
          <div
            class="main__data"
          >
            <div
              class="title"
            >
              testTitle
            </div>
            <div
              class="rating__container"
            >
              <div
                class="rating"
              >
                0
              </div>
            </div>
          </div>
          <div
            class="genre"
          >
            testGenre1, testGenre2
          </div>
          <div
            class="time__data"
          >
            <div
              class="release "
            >
              2018
            </div>
            <div
              class="runtime"
            >
              10min
            </div>
          </div>
          <div
            class="overview"
          >
            testOverview
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="movie__detail__container"
    >
      <img
        alt="testTitle"
        class="poster__img"
        src="testPosterPath"
      />
      <div
        class="info"
      >
        <div
          class="main__data"
        >
          <div
            class="title"
          >
            testTitle
          </div>
          <div
            class="rating__container"
          >
            <div
              class="rating"
            >
              0
            </div>
          </div>
        </div>
        <div
          class="genre"
        >
          testGenre1, testGenre2
        </div>
        <div
          class="time__data"
        >
          <div
            class="release "
          >
            2018
          </div>
          <div
            class="runtime"
          >
            10min
          </div>
        </div>
        <div
          class="overview"
        >
          testOverview
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
